Use unittest instead of Nose as base

koterpillar opened the pull request:
See #108 for motivation.

koterpillar committed:
Don't use Nose to run the Aloe's own test suite

koterpillar committed:
WIP use unittest runner

koterpillar committed:
callable_type to disguise functions as types

koterpillar committed:
Basic feature running

koterpillar committed:
Remove manual output re-printing

koterpillar committed:
Load features from directories

koterpillar committed:
Replace Nose attribute plugin

koterpillar committed:
Ensure reproducible order of feature running

koterpillar committed:
Fix running callbacks and discovery

koterpillar committed:
Fix tests for the new runner interface

koterpillar committed:
Wire up the verbose output

koterpillar committed:
Fix the executable entry point

koterpillar committed:
Remove references to Nose

koterpillar committed:
Remove Python 3-only syntax

koterpillar committed:
Fix syntax error handling in features

koterpillar committed:
Defer step loading until the tests are collected

koterpillar committed:
Remove Nose's assert_equal

koterpillar committed:
Don't depend on unittest's _FailedTest

koterpillar committed:
Update smoke test for unittest 3.4

koterpillar commented:
This finally passes on Python 3, but is still likely to be hard to make work properly on Python 2.

koterpillar committed:
Remove NOSE_GHERKIN_CLASS

koterpillar committed:
Use PY2 instead of PY3 for future Python version compatibility

koterpillar committed:
Fix parsing arguments on Python 2

koterpillar committed:
Fix code style warnings on Python 2

koterpillar committed:
Return proper type from __str__ and shortDescription on PY2

koterpillar committed:
Fix failed import test cases created by the loader

koterpillar committed:
Don't check for strings in loadTestsFromName

koterpillar committed:
Fix test discovery on Python 2

koterpillar committed:
Internalize Nose importer

koterpillar commented:
Passes on Python 2 too. `--no-ignore-python` is broken, and Django and Webdriver need an update too.

koterpillar committed:
Fix running Python tests as well as Gherkin features

koterpillar committed:
Rename base test case environment variable

koterpillar committed:
Move TestProgram to a separate module

koterpillar committed:
Fixes for converting aloe_django

danni commented on aloe/loader.py:
```diff
@@ -0,0 +1,199 @@
+Loader for tests written in Gherkin.
+"""
+
+from __future__ import unicode_literals
```
Single line?

danni commented on aloe/loader.py:
```diff
@@ -0,0 +1,199 @@
+            return
+
+        self.feature_dirs = [
+            dir_
```
Identity comprehension? Could this just be `list()`?

danni commented on aloe/loader.py:
```diff
@@ -0,0 +1,199 @@
+
+        if hasattr(self, 'after_hook'):
+            self.after_hook()
+            delattr(self, 'after_hook')
```
This feels weird, why not re-retrieve it from the registry like in `run_before_callbacks`?

danni commented on aloe/main.py:
```diff
@@ -0,0 +1,195 @@
+Unittest main working with Gherkin tests.
+"""
+
+from __future__ import unicode_literals
```
Single line?

danni commented on tests/factoryboy_app/features/steps.py:
```diff
@@ -178,4 +177,5 @@ def check_types(self):
                         ('date', date)):
-        assert_equal(type(getattr(MyWeirdObject.ref, attr)), type_)
+        # pylint:disable=unidiomatic-typecheck
+        assert type(getattr(MyWeirdObject.ref, attr)) == type_
```
`is`

danni commented on aloe/fs.py:
```diff
@@ -37,6 +37,100 @@ def path_to_module_name(filename):
+            finally:
+                if fh:
+                    fh.close()
+                release_lock()
```
Put some space in here to make the flow control easier to read :)

koterpillar replied:
Lifted from Nose, I'd rather replace this altogether or use some library.

danni reviewed (commented):
Generally ++. Minor thoughts.

koterpillar committed:
Style fixes

- Use a single __future__ import
- Avoid unnecessary list comprehension
- Use 'is' for comparing types

koterpillar committed:
Simplify running 'all' level callbacks

danni commented:
Note to self: I've promised to test this against one of my projects.

koterpillar committed:
Merge branch 'master' into pure-unittest

koterpillar commented:
There's no one brave enough to try the change on a live project and share the results here. With the level of invasiveness required to make both nose and, sadly, unittest work, I am reluctant to merge the changes in (see also https://github.com/aloetesting/aloe_django/pull/50) without such a confirmation.

kingbuzzman commented:
@koterpillar im not going to lie -- yes its a little scary, specially since i just finished creating a migration aloe testing implementation (im going to release it soon) and it heavily relies on nose. I'm going to test test when i get a chance, our code should be the perfect example:

```shell
$ find /code -type f -path '*/features/*' -name '*.py' -exec cat {} \; -exec echo \; | wc -l
   17630
$ find /code -type f -path '*/features/*' -name '*.feature' -exec cat {} \; -exec echo \; | wc -l
   88730
```

koterpillar commented:
Uncluttering my PRs.
